src/ui/VirtualFileRenderer/VirtualFileRenderer.spec.tsx

```diff
@@ -0,0 +1,361 @@
+import { fireEvent, render, screen, waitFor } from '@testing-library/react'
+import { userEvent } from '@testing-library/user-event'
+// eslint-disable-next-line no-restricted-imports
+import { type Dictionary } from 'lodash'
+import { MemoryRouter, Route, useLocation } from 'react-router-dom'
+
+import { VirtualFileRenderer } from './VirtualFileRenderer'
+
+window.requestAnimationFrame = (cb) => {
+  cb(1)
+  return 1
+}
+window.cancelAnimationFrame = () => {}
+
+const scrollToMock = jest.fn()
+window.scrollTo = scrollToMock
+
+class ResizeObserver {
+  callback = (x: any) => null
+
+  constructor(callback: any) {
+    this.callback = callback
+  }
+
+  observe() {
+    this.callback([{ target: { getAttribute: () => ({ scrollWidth: 100 }) } }])
+  }
+  unobserve() {
+    // do nothing
+  }
+  disconnect() {
+    // do nothing
+  }
+}
+global.window.ResizeObserver = ResizeObserver
+
+const code = `<Breadcrumb
+    paths={[
+    { pageName: 'owner', text: owner },
+    { pageName: 'repo', text: repo },
+    ...treePaths,
+    {..props}
+    ]}
+/>`
+
+const coverageData = {
+  0: 'H',
+  1: 'M',
+  2: 'P',
+  3: 'H',
+  4: 'M',
+  5: 'P',
+} as unknown as Dictionary<'H' | 'M' | 'P'>
+
+let testLocation: ReturnType<typeof useLocation>
+const wrapper =
+  (initialEntry = '/'): React.FC<React.PropsWithChildren> =>
+  ({ children }) => (
+    <MemoryRouter initialEntries={[initialEntry]}>
+      <Route path="/">{children}</Route>
+      <Route
+        path="*"
+        render={({ location }) => {
+          testLocation = location
+          return null
+        }}
+      />
+    </MemoryRouter>
+  )
```

Comment:
:wat:?

Reply:
So this is to wrap the tests, because we're looking to test for when the user clicks on the line number it updates the URL that requires this wrapper so that we have browser history present for the test.

Reply:
Hmm, and is there no other way to access or listen to location? It seems weird to have to render and assign to a local variable to escape this, surely there is a better way?

Reply:
So this is the way outlined in the "Testing" section for React-Router V5: https://v5.reactrouter.com/web/guides/testing/checking-location-in-tests

Reply:
Cant argue with official docs :)